Feature/open dyslexic font, fixes #115 (WIP)

danielweck opened the pull request:
#### This pull request is a Work In Progress

Demo app deployment URL: http://readium-fonts.surge.sh
#### Related issue(s) and/or pull request(s)
- https://github.com/readium/readium-shared-js/pull/298
- https://github.com/readium/readium-shared-js/issues/73
- https://github.com/readium/readium-js-viewer/issues/115
### Additional information

Currently users define font objects (see readium/readium-shared-js/pull/298 for more details.
Font objects get defined in readium-js-viewer through module config. Requirejs config defines a fonts array of font object which gets pulled into the readium-shared-js interface.
The UI currently works and is fairly finalized. hooking this into shared-js is still a WIP.

danielweck commented:
Test URL for deployed cloud reader app:
https://readium-fonts.surge.sh/?epub=https%3A%2F%2Freadium.firebaseapp.com%2Fepub_content%2Faccessible_epub_3

PROBLEM: FontLoader does not detect applied OpenDyslexic font-face / font-family (probably a Readium-specific issue due to the asynchronous nature of font loading combined with out-of-sync rendering flow), so the pagination is incorrect (OpenDyslexic metrics mean that there are more pages than normal, and Readium misses out on the last pages of a HTML content document)

danielweck commented on src/fonts/OpenDyslexic/OpenDyslexic/OpenDyslexic-Bold.svg:
```diff
@@ -0,0 +1,244 @@
+<?xml version="1.0" standalone="no"?>
+<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd" >
```
Hello @derekriemer in this commit there are duplicated font files (inside a new OpenDyslexic subfolder) in addition to src/js/ReaderSettingsDialog.js

derekriemer replied:
Oh, that's interesting. I'll remove these.


On 7/14/2016 1:48 AM, Daniel Weck wrote:
>
> In src/fonts/OpenDyslexic/OpenDyslexic/OpenDyslexic-Bold.svg 
> <https://github.com/readium/readium-js-viewer/pull/540#discussion_r70762812>:
>
> > @@ -0,0 +1,244 @@
> > +<?xml version="1.0" standalone="no"?>
> > +<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd" >
>
> Hello @derekriemer <https://github.com/derekriemer> in this commit 
> there are duplicated font files (inside a new OpenDyslexic subfolder) 
> in addition to src/js/ReaderSettingsDialog.js
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/readium/readium-js-viewer/pull/540/files/41229f6bade6f63f04851ed68233d9b5a76478b3..642bf5d9a886bdb2e765d61e05ee366db8b5a0f4#r70762812>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AFGivWy9_e6ugJhPTNH8pnWAXFR9rY5Jks5qVem4gaJpZM4I8jm9>.
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.com>
Awesome little hand built weather app! 
<http://django.derekriemer.com/weather/>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

danielweck commented on src/js/ModuleConfig.js:
```diff
@@ -1,6 +1,11 @@
 define(['module'], function(module) {
 
         var config = module.config();
+        console.log(config);
```
Note: let's remove this console logging message.

danielweck commented on src/js/ReaderSettingsDialog.js:
```diff
@@ -1,3 +1,4 @@
+var foo=5;
```
Can the `foo` var be removed? (it looks unused)

derekriemer replied:
Debug code that didn't get removed, I'll remove this.


On 7/19/2016 10:29 AM, Daniel Weck wrote:
>
> In src/js/ReaderSettingsDialog.js 
> <https://github.com/readium/readium-js-viewer/pull/540#discussion_r71374286>:
>
> > @@ -1,3 +1,4 @@
> > +var foo=5;
>
> Can the |foo| var be removed? (it looks unused)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/readium/readium-js-viewer/pull/540/files/41229f6bade6f63f04851ed68233d9b5a76478b3..642bf5d9a886bdb2e765d61e05ee366db8b5a0f4#r71374286>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFGivWjZRlpqwGZfVm43fz9eiwqrPOsZks5qXPuBgaJpZM4I8jm9>.
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.com>
Awesome little hand built weather app! 
<http://django.derekriemer.com/weather/>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

derekriemer replied:
Ready for more code review.

Thanks.


On 7/19/2016 10:29 AM, Daniel Weck wrote:
>
> In src/js/ReaderSettingsDialog.js 
> <https://github.com/readium/readium-js-viewer/pull/540#discussion_r71374286>:
>
> > @@ -1,3 +1,4 @@
> > +var foo=5;
>
> Can the |foo| var be removed? (it looks unused)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/readium/readium-js-viewer/pull/540/files/41229f6bade6f63f04851ed68233d9b5a76478b3..642bf5d9a886bdb2e765d61e05ee366db8b5a0f4#r71374286>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFGivWjZRlpqwGZfVm43fz9eiwqrPOsZks5qXPuBgaJpZM4I8jm9>.
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.com>
Awesome little hand built weather app! 
<http://django.derekriemer.com/weather/>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

danielweck commented on src/i18n/_locales/en_US/messages.json:
```diff
@@ -246,6 +246,23 @@
+  },
+    "i18n_fonttype-odf" : {
+    "message" : "Open Dyslexic Font"
+  },
```
TODO: this should not be hard-coded, instead the message string should be picked-up from `readium_js_viewer/ModuleConfig.fonts[n].displayName`

danielweck commented on src/i18n/_locales/en_US/messages.json:
```diff
@@ -246,6 +246,23 @@
+    "message" : "Font Testing Ground"
+  },
+    "i18n_font_test" : {
+    "message" : "This is a test of the font you picked. If you like it, then it is the font for you to read with."
```
TODO: decrease the verbosity of this l18n string (too hard to translate), re-word to make clearer.

derekriemer commented:
cc @clapierre @benetech

danielweck commented:
@olivierkorner I deployed the latest build of this PR to:
https://readium-fonts.surge.sh/?epub=https%3A%2F%2Freadium.firebaseapp.com%2Fepub_content%2Faccessible_epub_3

Please give this a try by navigating to subsequent pages (all the way into the next spine items), and back (using the default authored font). Then try again after disabling the browser cache (personally, I just open my web inspector which is configured to disable cache automatically). Notice the page "flicker", sometimes with noticeable display of a "compressed" column layout on the first page, immediately followed by the correct rendering of the last page of the reflowable document (when switching back from the beginning of the next spine item).
Then choose the OpenDyslexic font in the settings dialog, and try all of this again.
In the console log, I notice that the ResizeSensor triggers several times in a row (usually twice) even though the body dimensions remain the same. We refresh updatePagination() at every call, so this is better avoided :)
Otherwise, everything seems to be working pretty well (better than with FontLoader).
Dan

olivierkorner commented:
@danielweck Indeed the flicker is very noticeable with the OpenDyslexic font, less so but still present on some OS / browser. Maybe the pagination should be defer a little when the page loads to avoid this annoying flicker.

I also noticed that the ResizeSensor is called more than twice in a row with this setting, but only twice without it. When does the OpenDyslexic font take place? It looks like it's taking place in `UpdateHtmlFontAttributes`, which is called after at least one page layout if memory serves well.
To avoid this, shouldn't the OpenDyslexic stylesheet be inserted in the document as soon as the DOM is ready, rather than at the end of `onIFrameLoad`, with the update of the font size. I know it's not the most elegant solution... 

I don't see the "compressed" column layout you mentioned though. What is your testing platform and browser? I'm on a macOS, tested in Chrome, Safari and Firefox and also on Windows through Parallels Desktop.

I also see an error in triggerLayout due to insertRule:
on Firefox:
```
DOMException [SecurityError: "The operation is insecure."
code: 18
nsresult: 0x80530012
location: https://readium-fonts.surge.sh/scripts/readium-js-viewer_all.js:8] readium-js-viewer_all.js:8:4002

DOMException [SecurityError: "The operation is insecure."
code: 18
nsresult: 0x80530012
location: https://readium-fonts.surge.sh/scripts/readium-js-viewer_all.js:8]
```

on Safari:
```
Error: HierarchyRequestError: DOM Exception 3
insertRule
triggerLayout — readium-js-viewer_all.js:7465:84
C — readium-js-viewer_all.js:14661
applyStyles — readium-js-viewer_all.js:14723
(fonction anonyme) — readium-js-viewer_all.js:14575
h — readium-js-viewer_all.js:7392
(fonction anonyme) — readium-js-viewer_all.js:4313
```

but not on Chrome, curiously...

olivierkorner commented:
@danielweck in ReflowableView, the call to updatePagination in the ResizeSensor is not debounced like it is in ScrollView. In light of the burst of resize events, especially on Chrome (more than 10 in a rapid succession !), it should be debounced.
Is your test built commited somewhere? I'd like to have a look at this if you don't mind.

danielweck commented:
@olivierkorner  By test build do you mean 
https://readium-fonts.surge.sh/?epub=https%3A%2F%2Freadium.firebaseapp.com%2Fepub_content%2Faccessible_epub_3
?

olivierkorner commented:
@danielweck Yes the one you deployed at this test URL

danielweck commented:
It's just a standard readiumjs-viewer build:
npm run prepare && npm run dist+sourcemap

olivierkorner commented:
Ah ok sorry, I hadn't seen the last commits of open-dyslexic-font on readium-shared-j!

danielweck commented:
Essential code diffs: 
https://github.com/readium/readium-js-viewer/compare/develop...feature/open-dyslexic-font
https://github.com/readium/readium-shared-js/compare/develop...feature/openDyslexicFont

samkeller commented:
changing Essential code diffs done! but without success.
I can change the font to opendyslexic and the preview is OK but all sample epubs font styles remain the same as before. 
does anyone change a book's default font?

danielweck commented:
have you compiled the correct branch of shared-js into your build?

samkeller commented:
I compiled develop version and then change differences of feature/openDyslexicFont into it. and recompile.
is any working demo available?

danielweck commented:
demo build
https://readium-fonts.surge.sh/?epub=https%3A%2F%2Freadium.firebaseapp.com%2Fepub_content%2Faccessible_epub_3

samkeller commented:
Thanks. and finally it works for me too. the problem occurred in merging process but a fresh clone was OK.

samkeller commented:
is it possible to add more custom fonts?

danielweck commented:
There's no "easy way", but if you look for the code diff in the readium-js-viewer PR, and search for occurrences of the OpenDyslexic font, you can tell how it's done .

danielweck commented:
related https://github.com/readium/readium-shared-js/issues/271

danielweck commented:
Ah, I noticed that using the web browser's inspector / debugger / console in order to disable the cache results in weird rendering / layout of reflowable documents when column-paginated. Such as the "compressed" text and some additional content flicker (switch from the previous page to the next). These sorts of problems do not occur when the developer tools window is closed. Worth knowing.

danielweck commented:
Update: I fixed a number of issues, notably:

font family selector now styled appropriately in the settings dialog
added external web fonts for testing more than default + open dyslexic choices

In Chrome on Window 10, I notice that the Open Dyslexic font face loading results in a weird flicker, which I cannot reproduce with the other 3 fonts (default in "Accessible EPUB3" ebook, and the other 2 web fonts from the options selector). Perhaps that's because of the way the font faces are organised?
https://github.com/readium/readium-js-viewer/tree/feature/open-dyslexic-font/src/fonts/OpenDyslexic

Test URL:
https://readium-fonts.surge.sh/?epub=https%3A%2F%2Freadium.firebaseapp.com%2Fepub_content%2Faccessible_epub_3

olivierkorner commented:
@danielweck I notice that the OpenDyslexic font is not cached (HTTP headers are `Cache-control: no-cache`) and is reloaded for each view (actually the server responds with a status 304, but still...), whereas the Google fonts are served with some appropriate caching headers (`Cache-control: max-age=3600`). Maybe this explains the flicker.

danielweck commented:
Right, Surge.sh configured with no-cache, I guess.
Not sure about Firebase (which is used to deploy the app build from the develop branch).

danielweck commented:
Fixes https://github.com/readium/readium-js-viewer/issues/115

danielweck commented:
Travis CI ongoing build: https://travis-ci.org/readium/readium-js-viewer/builds/181698709

danielweck commented:
Test URL:
https://readium.firebaseapp.com/?epub=epub_content%2Faccessible_epub_3&goto=%7B%22idref%22%3A%22id-id2645682%22%2C%22elementCfi%22%3A%22%2F4%2F2%5Bconclusion%5D%2F12%2F12%2F6%2C%2F1%3A297%2C%2F1%3A302%22%7D

(`goto` query parameter / "shared bookmark" targets the 'EPUBs' term in a chapter at the end of the ebook entitled "Accessible EPUB3")

Use the left/right arrow keys to test the transition from one spine item to the previous/next (there should be no visible "flicker"). Hit the 'o' key to open the options dialog, and select the OpenDyslexic font from the pull-down list. Save the preferences, and try navigating the pages once again. Also try with the online Google fonts. Refresh the browser to re-activate the `goto` URL (the term 'EPUBs' should temporarily highlight in yellow).

pesce88 commented:
> is it possible to add more custom fonts?

yes is possible
